Make event handler house keeping in moveNodeToNewDocument more efficient

rniwa opened the pull request:
#### d2548129f1780abac960e3a93eaa547ba11073b6
<pre>
Make event handler house keeping in moveNodeToNewDocument more efficient
<a href="https://bugs.webkit.org/show_bug.cgi?id=267264">https://bugs.webkit.org/show_bug.cgi?id=267264</a>

Reviewed by Chris Dumez.

Prior to this PR, the code to update the number of wheel, touch, and gesture events in
Node::moveNodeToNewDocument were inefficient because it traversed EventListenerMap for
each distinct wheel, touch, and gesture event type. This is O(kn) where k is the total
number of event types, and n is the total number of event listeners on a given Node.

This PR reimplements the same house keeping in O(n): It enumerates types of event
listeners added on &quot;this&quot; event target and their count, and uses this information to
count the total number of wheel, touch, and gesture events in a single iteration over
EventListenerMap.

* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::AccessibilityObject::hasTouchEventListener const):
* Source/WebCore/dom/EventListenerMap.h:
(WebCore::EventListenerMap::containsMatchingEventListener):
* Source/WebCore/dom/EventTarget.h:
(WebCore::EventTarget::enumerateEventListenerTypes):
(WebCore::EventTarget::containsMatchingEventListener):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::moveNodeToNewDocument):
* Source/WebCore/dom/make-event-names.py:

Canonical link: <a href="https://commits.webkit.org/272834@main">https://commits.webkit.org/272834@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3cd91e76cfbd4854fbdd93d920811aefcd14164

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33233 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12009 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35147 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35870 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/30256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34204 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14354 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9181 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29394 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33708 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10130 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29685 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8830 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8977 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29660 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37201 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30197 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30040 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35087 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/9102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7056 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32947 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10831 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/9685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4272 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9785 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/EventListenerMap.h:
```diff
@@ -74,6 +74,16 @@ class EventListenerMap {
     }
 
+    template <typename CallbackType>
+    bool containsMatchingEventListener(CallbackType callback)
```
Would be nice to mark this function as `const`.

rniwa replied:
Indeed this can be const.

cdumez commented on Source/WebCore/dom/EventListenerMap.h:
```diff
@@ -74,6 +74,16 @@ class EventListenerMap {
             callback(entry.first, entry.second.size());
     }
 
+    template <typename CallbackType>
```
I think we usually omit the space?

rniwa replied:
Hm... we seem to mix both styles. 907 with a space and 4089 without a space. I guess without a space is more common so I'd align with that.

cdumez commented on Source/WebCore/dom/EventTarget.h:
```diff
@@ -150,6 +150,21 @@ class EventTarget : public ScriptWrappable, public CanMakeWeakPtrWithBitField<Ev
+    }
+
+    template <typename CallbackType>
+    bool containsMatchingEventListener(CallbackType callback)
```
Should probably be `const`.

cdumez commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2220,26 +2220,36 @@ void Node::moveNodeToNewDocument(Document& oldDocument, Document& newDocument)
             newDocument.didAddEventListenersOfType(eventType, count);
         });
 
+        unsigned numWheelEventHandlers = 0;
```
Would be nice to avoid the abbreviation in the variable names here for `num`, either by using `numberOf` as prefix or by using `Count` as suffix.

rniwa replied:
I'm just moving this declaration. If we rename it, we'd have to touch more lines.

cdumez replied:
Ah ok. No biggie.

cdumez commented on Source/WebCore/dom/EventTarget.h:
```diff
@@ -150,6 +150,21 @@ class EventTarget : public ScriptWrappable, public CanMakeWeakPtrWithBitField<Ev
     }
 
+    template <typename CallbackType>
+    void enumerateEventListenerTypes(CallbackType callback)
```
Can this be `const`?

rniwa replied:
Indeed fixed along side the space removal in the line above it.

rniwa committed:
Make event handler house keeping in moveNodeToNewDocument more efficient
https://bugs.webkit.org/show_bug.cgi?id=267264

Reviewed by Chris Dumez.

Prior to this PR, the code to update the number of wheel, touch, and gesture events in
Node::moveNodeToNewDocument were inefficient because it traversed EventListenerMap for
each distinct wheel, touch, and gesture event type. This is O(kn) where k is the total
number of event types, and n is the total number of event listeners on a given Node.

This PR reimplements the same house keeping in O(n): It enumerates types of event
listeners added on "this" event target and their count, and uses this information to
count the total number of wheel, touch, and gesture events in a single iteration over
EventListenerMap.

* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::AccessibilityObject::hasTouchEventListener const):
* Source/WebCore/dom/EventListenerMap.h:
(WebCore::EventListenerMap::containsMatchingEventListener):
* Source/WebCore/dom/EventTarget.h:
(WebCore::EventTarget::enumerateEventListenerTypes):
(WebCore::EventTarget::containsMatchingEventListener):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::moveNodeToNewDocument):
* Source/WebCore/dom/make-event-names.py:

Canonical link: https://commits.webkit.org/272834@main
